src/matrix-utils.ts

```diff
@@ -190,8 +197,11 @@ export async function initClient(
 
   await client.initCrypto();
   client.setGlobalErrorOnUnknownDevices(false);
+  // Apply listener before starting the client.
+  // Otherwise, the first sync could resolve before the listener gets applied.
+  const syncPromise = waitForSync(client);
   await client.startClient();
-  await waitForSync(client);
+  await syncPromise;
```

Comment:
As I understand it, no sync will happen until `startClient` is called, and since `startClient` [runs syncs asynchronously](https://github.com/matrix-org/matrix-js-sdk/blob/93d96281fd4cc8fd47c398eadac8b553662c85ac/src/client.ts#L1540) & communicates sync completion only via `ClientEvent.Sync` events, the only way to check if a sync has completed is to add a listener on that event -- and the only way to prevent a race condition between the event being fired & the listener being added is to add the listener before calling `startClient`.

I also don't think it would help to check if the client is done syncing first, because 1) if there is a cached sync state, the client will run two syncs on startup, and 2) after the initial sync, the client will be polling `/sync` forever.  I think this makes it hard to reason about when the client is really "done syncing", and it's much more straightforward to just listen for an expected sequence of `ClientEvent.Sync` events on startup.

Reply:
a1105a8 adds some comments to explain this a bit better. Let me know if they need any improvements.